controller/src/main/java/io/pravega/controller/store/stream/tables/TableHelper.java

```diff
@@ -340,49 +364,65 @@ public static long getSizeTillStreamCut(final byte[] indexTable, final byte[] hi
-                                            final long timeStamp) {
+    public static Pair<byte[], byte[]> addNewSegmentsToSegmentTableAndIndex(final int startingSegmentNumber,
+                                                                            final int newEpoch,
+                                                                            final byte[] segmentIndex,
+                                                                            final byte[] segmentTable,
+                                                                            final List<AbstractMap.SimpleEntry<Double, Double>> newRanges,
+                                                                            final long timeStamp) {
+
+        // if startingSegmentNumber was already previously indexed, overwrite the index
+        int segmentIndexOffset = SegmentIndexRecord.readRecord(segmentIndex, startingSegmentNumber)
+                .map(index -> index.getIndexOffset()).orElse(segmentIndex.length);
         final ByteArrayOutputStream segmentStream = new ByteArrayOutputStream();
+        final ByteArrayOutputStream indexStream = new ByteArrayOutputStream();
+        indexStream.write(segmentIndex, 0, segmentIndexOffset);
         segmentStream.write(segmentTable);
-
-        writeSegmentsToSegmentTable(startingSegmentNumber, creationEpoch, newRanges, timeStamp, segmentStream);
-
-        return segmentStream.toByteArray();
+        writeToSegmentTableAndIndex(startingSegmentNumber, newEpoch, newRanges, timeStamp, segmentStream, indexStream);
+        return new ImmutablePair<>(indexStream.toByteArray(), segmentStream.toByteArray());
     }
 
-    private static void writeSegmentsToSegmentTable(int startingSegmentNumber, int creationEpoch, List<AbstractMap.SimpleEntry<Double, Double>> newRanges, long timeStamp, ByteArrayOutputStream segmentStream) {
+    @SneakyThrows
+    private static void writeToSegmentTableAndIndex(int startingSegmentNumber, int newEpoch,
+                                                    List<AbstractMap.SimpleEntry<Double, Double>> newRanges,
+                                                    long timeStamp,
+                                                    ByteArrayOutputStream segmentStream,
+                                                    ByteArrayOutputStream indexStream) {
         IntStream.range(0, newRanges.size())
                 .forEach(
-                        newRange -> {
+                        x -> {
                             try {
-                                segmentStream.write(new SegmentRecord(startingSegmentNumber + newRange,
-                                        creationEpoch, timeStamp, newRanges.get(newRange).getKey(), newRanges.get(newRange).getValue())
+                                int offset = segmentStream.size();
+                                segmentStream.write(new SegmentRecord(startingSegmentNumber + x,
```

Comment:
This is an excellent example of `toByteArray` being redundant. If you had returned `ArrayView`, you would have called `write(av.array(), av.arrayOffset(), av.length())` which would only copy the data once, vs allocating an array just so that you can copy it wholly.